src/ruby/puppet-server-lib/puppet/server/network/http/handler.rb

```diff
@@ -32,7 +34,12 @@ def path(request)
   end
 
   def body(request)
-    request["body"]
+    body = request["body"]
+    if body.java_kind_of?(InputStream)
+      body.to_io.read()
```

Comment:
Yeah, this topic has come up a few times before and it's definitely something we should be keeping an eye on, as it would be an obvious and potentially significant improvement... but I don't think we can tackle it right now.